### ISSUE TITLE
[GraphQL] Doctrine association security

### ISSUE THREAD

**Issue by DavidBennettUK** [api-platform/core#4169]
There are some issues with the way security works with GraphQL. This can make creating an API secure from traversal attacks difficult/impossible.

Here's a simple example, exposing `User` and `Document` resources with simple security - where only admins can query the full collection, and voters are used to check if the user is allowed query a single item.

```
/**
 * @ORM\Entity
 */
#[ApiResource(
    graphql: [
        'item_query' => ['security' => 'is_granted("VIEW", object)'],
        'collection_query' => ['security' => 'is_granted("ROLE_SUPER_ADMIN")'],
    ],
)]
class User
{
    ...

    /**
     * @ORM\ManyToMany(targetEntity=Document::class, mappedBy="viewers")
     */
    private Collection $viewableDocuments;
}
```

```
/**
 * @ORM\Entity
 */
#[ApiResource(
    graphql: [
        'item_query' => ['security' => 'is_granted("VIEW", object)'],
        'collection_query' => ['security' => 'is_granted("ROLE_SUPER_ADMIN")'],
    ],
)]
class Document
{
    ...

    /**
     * @ORM\ManyToMany(targetEntity=User::class, inversedBy="viewableDocuments")
     */
    private Collection $viewers;

    /**
     * @ORM\ManyToOne(targetEntity=User::class)
     */
    protected ?User $createdBy = null;
}
```

There are two bugs(?) with the above example:
1. When querying a `User` with documents (e.g. `user(id) { documents: { totalCount } }`), API Platform will attempt to run the `item_query` `security` from `Document`, but with the paginator as the `object` - performing `is_granted('VIEW', object)` on the paginator. This results in access denied, as there's no voter that handles `Paginator`. 
2. ~~When querying a `Document` with a null `createdBy` field (e.g. `documents(id) { createdBy: { name } }`), API Platform will use the `security` from `User` `item_query` - performing `is_granted('VIEW', null)`. This will always cause access denied, as there will be no voter set up to handle a null subject. The only voter that handles a null subject is `RoleVoter`/`RoleHierarchyVoter` but obviously these won't handle it, as it's not a role that's being checked here. This seems like a bug - security shouldn't be ran if the field is null.~~ This has actually been indirectly resolved in 2.6.3 here https://github.com/api-platform/core/pull/4092 - as this causes security to be skipped for null fields.

I currently work around both of these issues with a `NullVoter` and `PaginatorVoter` that always grants access. Unless I'm missing something, I don't think that API Platform should be running security at all on associations, as it doesn't really make sense to - the security may depend on the context of where that property is. The `item_query` and `collection_query` security should only be used for the actual exposed GraphQL queries. Instead, I think it makes more sense to secure the individual fields with the `@ApiProperty` `security` attribute - however this is currently not working with GraphQL (see https://github.com/api-platform/core/pull/4143).

In the above example, I would want to secure `User::$documents` so that only certain users could access that collection (admins and the user themself). At the moment, it's not possible to do this. While the `Document` resource collection is secured from being queried directly (`Query.document(id)` or `Query.documents()`), if you are able to query the user through GraphQL, then you will be able to query their list of documents - potentially gaining access to documents that you shouldn't be able to view (traversal attack). There are ways to 'solve' this, such as making a custom normalizer to check view access on every object being serialized, however it's not possible to grant access based on the context of the query path. 

TLDR - I think that security shouldn't be run on Doctrine associations, and that the `@ApiResource` `security` attribute should be fixed to work with GraphQL to be used for securing association fields. Thoughts?

**Comment by DavidBennettUK** [comment]
PR ready to go (I think!) that covers this https://github.com/api-platform/core/pull/4143
